packages/routes-gen/src/route.ts

```diff
@@ -1,28 +1,33 @@
-export function route<T extends string>(
-  path: T,
-  params?: Record<string, any>
-): T {
-  if (params) {
-    const segments = path.split(/\/+/).map((segment) => {
-      if (segment.startsWith(":")) {
-        const key = segment.replace(":", "").replace("?", "");
+export function route<T extends string>(path: T, params: Record<string, any> = {}, includeOptional?: boolean): T {
+  // If the path doesn't contain any optional segments or parameter placeholders, return the path as is.
+  if (!path.includes('?') && !path.includes(':')) {
+    return path;
+  }
 
-        if (key in params) {
-          return params[key];
-        }
+  let realPath = "";
+  let currentIndex = path.length;
+  let lastSegmentHadParam = false;
 
-        // If the segment is optional and it doesn't exist in params, return null to omit it from the resulting path
-        if (segment.endsWith("?")) {
-          return null;
-        }
+  while (currentIndex > 0) {
+    const startSegmentIndex = path.lastIndexOf('/', currentIndex);
+    const segment = path.slice(startSegmentIndex, currentIndex + 1);
+    currentIndex = startSegmentIndex - 1;
+    if (segment.startsWith('/:')) {
+      const paramName = segment.endsWith('?') ? segment.slice(2, -1) : segment.slice(2);
+      const paramValue = params[paramName];
+      if (paramValue !== undefined) {
+        lastSegmentHadParam = true;
+        realPath = `/${paramValue}` + realPath;
       }
-
-      return segment;
-    });
-
-    // Filter out any null/undefined segments and join remaining segments
-    return segments.filter((value) => value != null).join("/") as T;
+    } else if (segment.endsWith('?')) {
+      if (lastSegmentHadParam || includeOptional) {
+        realPath = segment.slice(0, -1) + realPath;
+      }
```

Comment:
this is where we handle non-param segments and the ability of opting into including non-param optional segments